text/0000-futures.md

```diff
@@ -154,31 +164,316 @@ can be dropped.
+  been driven to completion, all resources which had been allocated for the task
+  must have been released.
+- It must be safe to call `wake()` on a `Waker` even if the associated task has
+  already been driven to completion.
+- `Waker::wake()` must wake up an executor even if it is called from an arbitry
+  thread.
+
+An executor which implements `RawWaker` must therefore make sure that all these
+requirements are fulfilled.
+
+Since many the ownership semantics that are required here can easily be met
+through a reference-counted `Waker` implementation, a convienence method for
+defining `Waker`s is provided, which does not require implementing a `RawWaker`
+and the associated vtable manually.
+
+This convience method is based around the `ArcWake` trait. An implementor of
+an executor can define a type which implements the `ArcWake` trait as defined
+below. The `ArcWake` type defines the associated method `into_local_waker`,
+which allows to retrieve a `LocalWaker` instance from an `Arc` of this type.
+The returned instance will guarantee that the `wake()` and `wake_local` methods
+of the type which implements `ArcWake` are called, whenever `wake()` is called
+on a `Waker` or `LocalWaker`.
 
 ```rust
 /// A way of waking up a specific task.
 ///
-/// Any task executor must provide a way of signaling that a task it owns
-/// is ready to be `poll`ed again. Executors do so by providing a wakeup handle
-/// type that implements this trait.
-pub trait Wake: Send + Sync {
+/// By implementing this trait, types that are expected to be wrapped in an `Arc`
+/// can be converted into `LocalWaker` and `Waker` objects.
+/// Those Wakers can be used to signal executors that a task it owns
+/// is ready to be `poll`ed again.
+pub trait ArcWake: Send + Sync {
     /// Indicates that the associated task is ready to make progress and should
     /// be `poll`ed.
     ///
+    /// This function can called from the thread on which the `ArcWake` was created,
+    /// as well as from any other thread.
+    ///
     /// Executors generally maintain a queue of "ready" tasks; `wake` should place
     /// the associated task onto this queue.
     fn wake(self: &Arc<Self>);
 
     /// Indicates that the associated task is ready to make progress and should be polled.
-    /// This function is like wake, but can only be called from the thread on which this
-    /// `Wake` was created.
+    /// This function is like wake, but will only be called from the thread on which this
+    /// `ArcWake` was created.
+    ///
+    /// This function is marked unsafe because callers must guarantee that
+    /// they call it only on the task on which the `ArcWake` had been created.
     ///
     /// Executors generally maintain a queue of "ready" tasks; `wake_local` should place
     /// the associated task onto this queue.
-    unsafe fn wake_local(self: &Arc<Self>)
+    unsafe fn wake_local(self: &Arc<Self>);
```

Comment:
This should have a default impl to match the current one on nightly:

```rust
unsafe fn wake_local(arc_self: &Arc<Self>) {
    Self::wake(arc_self);
}
```

This makes it possible to create an entirely-safe `ArcWake`.

Reply:
Yes, I have done this always one my CR for actual rustc locally. Will add it.

However I have left a few comments around the usability of `ArcWake` in general in the comments of the CR. Please review them and comment on what you think.

Reply:
> in the comments of the CR

Sorry, I'm not sure which ones you're referring to, would you mind linking?

Reply:
https://github.com/aturon/rfcs/pull/15#issuecomment-440159309

In short: Most code seems to really want to create wakers starting through `LocalWakerRef` instead of the associated associated methods, but understanding this (and the whole local/nonlocal thing) seems to be pretty complicated. Therefore it's a bit questionable whether the associated method should be available on `ArcWake`, or whether we want to have `ArcWake` in std/alloc at all.

Reply:
I'm less concerned about `wake_local` and `wake` on the trait and more about the different ways to construct wakers from it. If most people end up not using `::into_local_waker(Self)` in favor of higher performance alternatives like `LocalWakerRef` then it seems a bit questionable. Especially since the audience for these kind of methods is very small (executor authors), and those might want to optimize for high performance.

Reply:
I think the overall number of people implementing any of this will be small. `LocalWakerRef` already exists and is usable. This API should exist to make it clear that this is possible. We could add `LocalWakerRef` to std if we really want at some point, but I don't think it matters greatly either way.